src/thor/costmatrix.cc

```diff
@@ -152,70 +150,87 @@ void CostMatrix::SourceToTarget(Api& request,
   SetSources(graphreader, source_location_list, time_infos);
   SetTargets(graphreader, target_location_list);
 
+  // Update hierarchy limits
+  if (!ignore_hierarchy_limits_)
+    ModifyHierarchyLimits();
+
   // Perform backward search from all target locations. Perform forward
   // search from all source locations. Connections between the 2 search
   // spaces is checked during the forward search.
-  int n = 0;
+  uint32_t n = 0;
   while (true) {
-    // Iterate all target locations in a backwards search
-    for (uint32_t i = 0; i < target_count_; i++) {
-      if (target_status_[i].threshold > 0) {
-        target_status_[i].threshold--;
-        BackwardSearch(i, graphreader);
-        // if we didn't see this
-        if (target_status_[i].threshold == 0) {
-          for (uint32_t source = 0; source < source_count_; source++) {
-            //  Get all targets remaining for the origin
-            auto& targets = source_status_[source].remaining_locations;
+    for (uint32_t i = 0; i < locs_count_[MATRIX_REV]; i++) {
+      if (locs_status_[MATRIX_REV][i].threshold > 0) {
+        locs_status_[MATRIX_REV][i].threshold--;
+        Expand<MatrixExpansionType::reverse>(i, n, graphreader);
+        // if we exhausted this search
+        if (locs_status_[MATRIX_REV][i].threshold == 0) {
+          for (uint32_t source = 0; source < locs_count_[MATRIX_FORW]; source++) {
+            // update the target for each source's remaining targets, if it still exists
+            auto& targets = locs_status_[MATRIX_FORW][source].unfound_connections;
```

Comment:
haha cool yeah there you did do the alias thing i was suggesting and it indeed saved some small diffs haha. i still cant really read this without manually comparing each line though